### PR TITLE
User permissions

### PR DESCRIPTION
**Pull request by smk4664**
# Closes: #112 
## What's Changed

Adds a new feature to Link the Nautobot User with a Chat User. This allows more granular permissions for chat commands, as the Chat User will only see devices/racks/circuits/etc. that they have permissions to in Nautobot. This will open up new possibilities for Chat Commands, like running jobs, creating Notes, creating/updating records etc.

## TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/chatops/en/latest/dev/contributing/#creating-changelog-fragments))
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [x] Documentation Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot_chatops/api/views/lookup.py [hunk 1]
```diff
@@ -32,3 +33,26 @@ def get(self, request, *args, **kwargs):
...
+class UserEmailLookupView(View):
+    """Look up a user_id by email."""
+
+    http_method_names = ["get"]
+
+    def get(self, request, *args, **kwargs):
+        """Handle an inbound GET request for a specific access grant value."""
+        for required_param in ("email", "platform"):
+            if required_param not in request.GET:
+                return HttpResponseBadRequest(f"Missing mandatory parameter {required_param}")
```
Comment (snaselj) [inline nautobot_chatops/api/views/lookup.py:47]:
```suggestion
        missing_params = set(("email", "platform")) - set(request.GET.keys())
        if missing_params:
            return HttpResponseBadRequest(f"Missing mandatory parameter {missing_params}")
```

The new version notifies user about both missing params.

File: nautobot_chatops/api/views/lookup.py [hunk 2]
```diff
@@ -32,3 +33,26 @@ def get(self, request, *args, **kwargs):
...
+        value = None
+        for dispatcher_class in Dispatcher.subclasses():
+            if dispatcher_class.platform_slug == request.GET["platform"]:
+                with contextlib.suppress(NotImplementedError):
+                    value = dispatcher_class.lookup_user_id_by_email(request.GET["email"])
+        return (
+            JsonResponse(data={"user_id": value})
+            if value
+            else HttpResponseNotFound(f"No user_id found for {request.GET['email']}")
+        )
```
Comment (snaselj) [inline nautobot_chatops/api/views/lookup.py:58]: Proposing to simplify this by starting with

```
email = request.GET["email"]
if not email:
    ...
platform = request.GET["platform"]
...
```

Code will be better readable IMHO

File: nautobot_chatops/api/views/lookup.py (hunk 1 above)
Comment (snaselj) [inline nautobot_chatops/api/views/lookup.py:47]:
```suggestion
        missing_params = set(["email", "platform"]) - set(request.GET.keys())
        if missing_params:
            return HttpResponseBadRequest(f"Missing mandatory parameter(s) {missing_params}")
```

File: nautobot_chatops/api/views/lookup.py (hunk 2 above)
Comment (snaselj) [inline nautobot_chatops/api/views/lookup.py:58]: Should not be `return` statement indented to the same level as `value = ...`? 

Or some `def find_dispatcher(name)` could be handy.

File: nautobot_chatops/models.py [hunk 3]
```diff
@@ -54,6 +56,13 @@ class CommandLog(PrimaryModel):
         default=CommandStatusChoices.STATUS_SUCCEEDED,
     )
     details = models.CharField(max_length=255, default="")
+    system_user = models.ForeignKey(
```
Comment (snaselj) [inline nautobot_chatops/models.py:59]: Would not be `nautobot_user` more descriptive name?

File: docs/models/chatopsaccountlink.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+Nautobot ChatOps now uses the built-in Nautobot permissions for Nautobot Objects (Devices, Locations, Racks, etc.). Each user will need to link their Nautobot Account with their Chat Platform User Account. Login to Nautobot then access the Link ChatOps Account within the Plugins menu. Here you can provide your email address and select the ChatOps Platform you are using, then click the Look up User ID from Email to get your Chat User ID.
+
+A Nautobot user can have multiple Chat users connected, but the Chat User can only be linked to one Nautobot user. As an example my team is transitioning from Slack to Mattermost. My Slack User ID and my Mattermost User ID can both be connected to the same Nautobot User.
+
+![Link Accounts](../images/account_link.png)
+
+## Configuring Account Link for many users
+
+Admins have the ability to access the Nautobot Admin page and can use `ChatOps Account Links` page to link multiple accounts, but you will need to know the Chat Platforms user ids for each user you are linking.
```
Comment (snaselj) [inline docs/models/chatopsaccountlink.md:14]:
```suggestion
Admins have the ability to access the Nautobot Admin page and can use `ChatOps Account Links` page to link multiple accounts, but you will need to know the Chat Platforms user IDs for each user you are linking.
```

File: nautobot_chatops/__init__.py [hunk 5]
```diff
@@ -54,6 +54,7 @@ class NautobotChatOpsConfig(PluginConfig):
         # Should menus, text input fields, etc. be deleted from the chat history after the user makes a selection?
         "delete_input_on_submission": False,
         "restrict_help": False,
+        "fallback_chatops_user": "chatbot",
```
Comment (snaselj) [inline nautobot_chatops/__init__.py:57]: Should not be added to development/nautobot_config.py?

File: nautobot_chatops/choices.py [hunk 6]
```diff
@@ -44,3 +44,19 @@ class CommandTokenPlatformChoices(ChoiceSet):
     WEBEX = "webex"
 
     CHOICES = ((MATTERMOST, "Mattermost"),)
+
+
+class PlatformChoices(ChoiceSet):
```
Comment (snaselj) [inline nautobot_chatops/choices.py:49]: Seems to me duplicit with CommandTokenPlatformChoices.

File: nautobot_chatops/dispatchers/base.py [hunk 7]
```diff
@@ -38,6 +41,26 @@ def __init__(self, context=None):
...
+                    platform=self.platform_slug, user_id=self.context["user_id"]
+                ).nautobot_user
+            except ObjectDoesNotExist:
+                logger.warning(
+                    f"Could not find User matching {self.context['user_name']} - id: {self.context['user_id']}."
+                    "Add a ChatOps User to link the accounts."
+                )
+        user_model = get_user_model()
+        user, _ = user_model.objects.get_or_create(
+            username=settings.PLUGINS_CONFIG["nautobot_chatops"]["fallback_chatops_user"]
```
Comment (snaselj) [inline nautobot_chatops/dispatchers/base.py:60]:
```suggestion
            username=_APP_CONFIG["fallback_chatops_user"]
```

File: nautobot_chatops/dispatchers/base.py [hunk 8]
```diff
@@ -38,6 +41,26 @@ def __init__(self, context=None):
...
+        )
```
Comment (snaselj) [inline nautobot_chatops/dispatchers/base.py:61]:
```suggestion
        user, _ = user_model.objects.get_or_create(username=_APP_CONFIG["fallback_chatops_user"])
```

File: nautobot_chatops/models.py [hunk 9]
```diff
@@ -142,7 +151,30 @@ class Meta:
         unique_together = [["platform", "token"]]
 
 
+class ChatOpsAccountLink(PrimaryModel):
```
Comment (snaselj) [inline nautobot_chatops/models.py:154]: `development/mattermost/nautobot_bootstrap.py` can be updated with value linkink nautobot admin to mattermost admin

**Review by snaselj — COMMENTED** [review]
Left a few suggestions, with my limited knowledge LGTM

Great feaure :+1:

File: nautobot_chatops/choices.py (hunk 6 above)
Comment (smk4664) [reply nautobot_chatops/choices.py:49]: Yeah, I realize that now, I think in my mind I was going to replace CommandTokenPlatformChoices with PlatformChoices.

File: nautobot_chatops/api/views/lookup.py (hunk 2 above)
Comment (smk4664) [reply nautobot_chatops/api/views/lookup.py:58]: It needs to return either way, this was recommended over having multiple return statements.

**Pull request merged**